pkg/controller/seed-controller-manager/operating-system-profile-synchronizer/controller.go

```diff
@@ -0,0 +1,289 @@
+	}
+	return r.syncAllUserClusterNamespaces(ctx, log, osp)
+}
+
+func (r *Reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, osp *osmv1alpha1.OperatingSystemProfile) error {
+	err := r.syncAllUserClusterNamespaces(ctx, log, osp)
+	if err != nil {
+		return err
+	}
+
+	// Remove the finalizer
+	return kuberneteshelper.TryRemoveFinalizer(ctx, r.seedClient, osp, cleanupFinalizer)
+}
+
+func (r *Reconciler) syncAllUserClusterNamespaces(ctx context.Context, log *zap.SugaredLogger, osp *osmv1alpha1.OperatingSystemProfile) error {
+	clusters := &kubermaticv1.ClusterList{}
+	if err := r.seedClient.List(ctx, clusters); err != nil {
+		log.Error(err)
+		utilruntime.HandleError(fmt.Errorf("failed to list clusters: %w", err))
+	}
+
+	var errors []error
+	for _, cluster := range clusters.Items {
+		// Ensure that this is a reconcilable cluster
+		if cluster.Spec.EnableOperatingSystemManager && cluster.DeletionTimestamp == nil && !cluster.Spec.Pause {
+			// We want to avoid changes to the original object.
+			currentOSP := osp.DeepCopy()
+			err := r.syncOperatingSystemProfile(ctx, log, currentOSP, cluster.Status.NamespaceName)
+			if err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+	return kerrors.NewAggregate(errors)
+}
+
+func (r *Reconciler) syncOperatingSystemProfile(ctx context.Context, log *zap.SugaredLogger, osp *osmv1alpha1.OperatingSystemProfile, namespace string) error {
+	// We want to perform actions on the user-cluster namespace.
+	osp.Namespace = namespace
+
+	// If OSP is marked for deletion then remove it from the user-cluster namespace.
+	if osp.DeletionTimestamp != nil {
+		err := r.seedClient.Delete(ctx, osp)
+		return ctrlruntimeclient.IgnoreNotFound(err)
+	}
+
+	existingOSP := &osmv1alpha1.OperatingSystemProfile{}
+	err := r.seedClient.Get(ctx, types.NamespacedName{Name: osp.Name, Namespace: osp.Namespace}, existingOSP)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get operatingSystemProfile %s/%s: %w", osp.Namespace, osp.Name, err)
+	}
+
+	// We need to check if the existing OperatingSystemProfile can be updated.
+	// OSP is immutable by nature and to make modifications a version bump is mandatory
+	if equal := apiequality.Semantic.DeepEqual(existingOSP.Spec, osp.Spec); !equal && existingOSP.Spec.Version == osp.Spec.Version {
+		log.Debug("OperatingSystemProfile is immutable. For updates .spec.version needs to be updated")
+		return nil
+	}
+
+	// We need to update OSP
+	if existingOSP.Name == "" {
+		// set resource version to empty.
+		osp.ResourceVersion = ""
+		osp.Finalizers = nil
+	}
+
+	var ospCreators []reconciling.NamedOperatingSystemProfileCreatorGetter
+	ospCreators = append(ospCreators, ospCreator(osp.Name, osp))
+
+	if err := osmreconciling.ReconcileOperatingSystemProfiles(ctx,
+		ospCreators,
+		namespace, r.seedClient); err != nil {
+		return fmt.Errorf("failed to reconcile osps: %w", err)
+	}
+	return nil
+}
```

Comment:
@xrstf I think I'm not understanding this correctly because this is a very shitty implementation from my end. Just let me know how it should have been refactored and I'll update it.